File: AlgoExpert/1. Easy/Python/11. Insertion Sort/Insertion Sort.py
# Best: O(n) time | O(1) space
# Average: O(n^2) time | O(1) space
# Worst: O(n^2) time | O(1) space

def insertionSort(array):
	for i in range(1, len(array)):
		j = i
		while j > 0 and array[j] < array[j - 1]:
			swap(j, j - 1, array)
			j -= 1
	return array

def swap(i, j, array):
	array[i], array[j] = array[j], array[i]

File: AlgoExpert/1. Easy/C++/11. Insertion Sort/Insertion Sort.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> insertionSort(vector<int> array);

// Best: O(n) time | O(1) space
// Average: O(n^2) time | O(1) space
// Worst: O(n^2) time | O(1) space

vector<int> insertionSort(vector<int> array) {
	if (array.empty()) {
		return {};
	}
	for (int i = 1; i < array.size(); i++) {
		int j = i;
		while (j > 0 && array[j] < array[j - 1]) {
			swap(array[j], array[j - 1]);
			j -= 1;
		}
	}
	return array;
}